catchup/peerSelector.go

```diff
@@ -0,0 +1,287 @@
+	return true
+}
+
+// sort the pools array in an accending order according to the rank of each pool.
+func (ps *peerSelector) sort() {
+	sort.SliceStable(ps.pools, func(i, j int) bool {
+		return ps.pools[i].rank < ps.pools[j].rank
+	})
+}
+
+// peerAddress returns the peer's underlying address. The network.Peer object cannot be compared
+// to itself, since the network package dynamically creating a new instance on every network.GetPeers() call.
+// The method retrun the peer address or an empty string if the peer is not one of HTTPPeer/UnicastPeer
+func peerAddress(peer network.Peer) string {
+	if httpPeer, ok := peer.(network.HTTPPeer); ok {
+		return httpPeer.GetAddress()
+	} else if unicastPeer, ok := peer.(network.UnicastPeer); ok {
+		return unicastPeer.GetAddress()
+	}
+	return ""
+}
+
+// refreshAvailablePeers reload the available peers from the network package, add new peers along with their
+// corresponding initial rank, and deletes peers that have been dropped by the network package.
+func (ps *peerSelector) refreshAvailablePeers() {
+	existingPeers := make(map[string]network.Peer)
+	for _, pool := range ps.pools {
+		for _, localPeer := range pool.peers {
+			if peerAddress := peerAddress(localPeer.peer); peerAddress != "" {
+				existingPeers[peerAddress] = localPeer.peer
+			}
+		}
+	}
+	sortNeeded := false
+	for _, initClass := range ps.peerClasses {
+		peers := ps.net.GetPeers(initClass.peerClass)
+		for _, peer := range peers {
+			peerAddress := peerAddress(peer)
+			if peerAddress == "" {
+				continue
+			}
+			if _, has := existingPeers[peerAddress]; has {
+				delete(existingPeers, peerAddress)
+				continue
+			}
+			// it's an entry which we did not had before.
+			sortNeeded = ps.addToPool(peer, initClass.initialRank, initClass) || sortNeeded
+		}
+	}
+
+	// delete from the pools array the peers that do not exist on the network anymore.
+	for poolIdx := len(ps.pools) - 1; poolIdx >= 0; poolIdx-- {
+		pool := ps.pools[poolIdx]
+		for peerIdx := len(pool.peers) - 1; peerIdx >= 0; peerIdx-- {
+			peer := pool.peers[peerIdx].peer
+			if peerAddress := peerAddress(peer); peerAddress != "" {
+				if _, has := existingPeers[peerAddress]; has {
+					// need to be removed.
+					pool.peers = append(pool.peers[:peerIdx], pool.peers[peerIdx+1:]...)
+				}
+			}
+		}
+		if len(pool.peers) == 0 {
+			ps.pools = append(ps.pools[:poolIdx], ps.pools[poolIdx+1:]...)
+			sortNeeded = true
+		} else {
+			ps.pools[poolIdx] = pool
+		}
+	}
+
+	if sortNeeded {
+		ps.sort()
+	}
+}
+
+// findPeer look into the peer pool and find the given peer.
+// The method returns the pool and peer indices if a peer was found, or (-1, -1) otherwise.
+func (ps *peerSelector) findPeer(peer network.Peer) (poolIdx, peerIdx int) {
+	peerAddr := peerAddress(peer)
+	if peerAddr != "" {
+		for i, pool := range ps.pools {
+			for j, localPeerEntry := range pool.peers {
```

Comment:
This is called every time RankPeer is called.
Why not keep a map from address to rank? Then, knowing the rank of the address, can do binary search on pools (since it is sorted), and locate the pool where the address/peer resides. 
refreshAvailablePeers, RankPeer, PeerDownloadDurationToRank, findPeer and GetNextPeer will all benefit from such a data structure, and is not costly to maintain it.

Reply:
I think that we can definitely should consider this in the future, if we see that this operation takes notable amount of time.

At this time, we don't have enough relays to really justify that. The good part is that if we do want to optimize it, the entire optimization can be done within this file.